### PR TITLE
Add reconfiguration engine to concord bft

### PR DESCRIPTION
**Pull request by yontyon**
Integrate reconfiguration engine into concordbft

Concordbft has its won reconfiguration engine that supports the following:
1. wedge 
2. pruning
Yet, the user may provide additional handlers for both, wedge and pruning in the command handler.

Meaning, the reconfiguration dispatcher holds lists of reconfiguration and pruning handlers. For every request in cmf, the dispatcher calls to all of the handlers one after another. The execution result is the logic & among all handlers.

This way, we have a default set of reconfiguration actions in concordbft + user-defined actions.

In addition, this PR also contains apollo tests for the new reconfiguration engine

### CODE REVIEW

File: bftclient/include/bftclient/base_types.h [hunk 1]
```diff
@@ -36,7 +36,13 @@ struct ClientId {
   bool operator<(const ClientId& other) const { return val < other.val; }
 };
 
-enum Flags : uint8_t { EMPTY_FLAGS_REQ = 0x0, READ_ONLY_REQ = 0x1, PRE_PROCESS_REQ = 0x2, KEY_EXCHANGE_REQ = 0x8 };
+enum Flags : uint8_t {
+  EMPTY_FLAGS_REQ = 0x0,
+  READ_ONLY_REQ = 0x1,
+  PRE_PROCESS_REQ = 0x2,
+  KEY_EXCHANGE_REQ = 0x8,
+  RECONFIG_FLAG = 0x20
```
Comment (salieri11) [inline bftclient/include/bftclient/base_types.h:44]: why not 0x10?

File: bftengine/include/bftengine/Replica.hpp [hunk 2]
```diff
@@ -50,20 +55,28 @@ enum MsgFlag : uint8_t {
...
+  virtual ~IControlHandler() = default;
 };
 
 class IRequestsHandler {
  public:
   struct ExecutionRequest {
     uint16_t clientId = 0;
     uint64_t executionSequenceNum = 0;
-    uint8_t flags = 0;
+    uint8_t flags = 0;  // copy of ClientRequestMsg flags
```
Comment (salieri11) [inline bftengine/include/bftengine/Replica.hpp:79]: I would use here the enum type to make it explicit

File: bftengine/src/bftengine/BFTEngine.cpp [hunk 3]
```diff
@@ -112,7 +112,7 @@ void ReplicaInternal::restartForDebug(uint32_t delayMillis) {
 namespace bftEngine {
 
 IReplica::IReplicaPtr IReplica::createNewReplica(const ReplicaConfig &replicaConfig,
-                                                 IRequestsHandler *requestsHandler,
+                                                 shared_ptr<IRequestsHandler> requestsHandler,
```
Comment (salieri11) [inline bftengine/src/bftengine/BFTEngine.cpp:115]: Doesn't it make sense here to use unique_ptr? I think that request handler will never have shared ownership?

File: bftengine/include/bftengine/Replica.hpp [hunk 4]
```diff
@@ -50,20 +55,28 @@ enum MsgFlag : uint8_t {
 // 2. Key exchange
 // 3. Change DB scheme
 // and basically any management action that is handled by the layer that uses concord-bft.
-class ControlHandlers {
+class IControlHandler {
  public:
+  static const std::shared_ptr<IControlHandler> instance(IControlHandler *ch = nullptr) {
```
Comment (robem) [inline bftengine/include/bftengine/Replica.hpp:60]: Why are we applying a singleton pattern here? Can't we instantiate one ControlHandler and pass around its reference?

File: bftengine/src/bftengine/ControlHandler.hpp [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  void onSuperStableCheckpoint() override {
+    onNoutOfNCheckpoint = true;
+    for (auto& cb : onNOutOfNCheckpointCallBack) cb();
+  };
+  void onStableCheckpoint() override { onNMinusFOutOfNCheckpoint = true; }
+  bool onPruningProcess() override { return onPruningProcess_; }
+  bool isOnNOutOfNCheckpoint() const override { return onNoutOfNCheckpoint; }
+  bool isOnStableCheckpoint() const override { return onNMinusFOutOfNCheckpoint; }
+  void setOnPruningProcess(bool inProcess) override { onPruningProcess_ = inProcess; }
+  void addOnNOutOfNCheckpointCallBack(const std::function<void()>& cb) override {
```
Comment (robem) [inline bftengine/src/bftengine/ControlHandler.hpp:31]: `NOutOfN` means super stable, correct? If so, then let's be consistent with our naming. We use `onSuperStableCheckpoint` above, let's use `addSuperStableCheckpointCallback` here.
The same applies for the variable names below.

File: bftengine/include/bftengine/Replica.hpp [hunk 6]
```diff
@@ -50,20 +55,28 @@ enum MsgFlag : uint8_t {
 // 2. Key exchange
 // 3. Change DB scheme
 // and basically any management action that is handled by the layer that uses concord-bft.
-class ControlHandlers {
+class IControlHandler {
```
Comment (robem) [inline bftengine/include/bftengine/Replica.hpp:58]: We have a separate `ControlHandler.hpp`. Would you mind adding this class to the same file or create a new one?

Also, do we need this interface class? Why don't we just use a ControlHandler? Do we foresee in the near future that we have multiple ControlHandler implementations?

File: bftengine/src/bftengine/ControlHandler.hpp [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+// file.
+
+#pragma once
+
+#include "Replica.hpp"
+#include <functional>
+
+namespace bftEngine {
+
+class ControlHandler : public IControlHandler {
```
Comment (robem) [inline bftengine/src/bftengine/ControlHandler.hpp:20]: Please add documentation so that it becomes clear what the ControlHandler does. I saw that there is documentation for the IControlHandler class, maybe move it into this file?

File: bftengine/src/bftengine/ControlHandler.hpp [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  bool onPruningProcess() override { return onPruningProcess_; }
+  bool isOnNOutOfNCheckpoint() const override { return onNoutOfNCheckpoint; }
+  bool isOnStableCheckpoint() const override { return onNMinusFOutOfNCheckpoint; }
+  void setOnPruningProcess(bool inProcess) override { onPruningProcess_ = inProcess; }
+  void addOnNOutOfNCheckpointCallBack(const std::function<void()>& cb) override {
+    onNOutOfNCheckpointCallBack.emplace_back(cb);
+  }
+
+ private:
+  bool onNoutOfNCheckpoint = false;
```
Comment (robem) [inline bftengine/src/bftengine/ControlHandler.hpp:36]: Private variables should end with `_`

File: bftengine/src/bftengine/ReplicaBase.hpp [hunk 9]
```diff
@@ -12,7 +12,7 @@
 #pragma once
 
 #include <memory>
-
+#include "Replica.hpp"
```
Comment (robem) [inline bftengine/src/bftengine/ReplicaBase.hpp:15]: This seems wrong. Why would a ReplicaBase include a Replica? Shouldn't it be the other way around?
It looks like you need `IRequestsHandler`. Why don't we introduce an `IRequestsHandlers.hpp` instead?

File: bftengine/src/bftengine/ReplicaImp.hpp [hunk 10]
```diff
@@ -12,7 +12,6 @@
 #pragma once
 
 #include <string>
-#include "Replica.hpp"
```
Comment (robem) [inline bftengine/src/bftengine/ReplicaImp.hpp:15]: Why did you remove `Replica.hpp` - `IRequestsHandler` is used in here and it is defined in the aforementioned header.

File: messages/compiler/cmfc.py [hunk 11]
```diff
@@ -35,6 +35,7 @@ def translate(ast, language, namespace, output):
         print("Generating Python source code")
         from python import pygen
         code = pygen.translate(ast, namespace)
+        print(output)
```
Comment (robem) [inline messages/compiler/cmfc.py:38]: That seems arbitrary?! Why do we print here? Also, there is a print statement right above.

**Review by robem — CHANGES_REQUESTED** [review]

File: tests/apollo/test_skvbc_reconfiguration.py [hunk 12]
```diff
@@ -59,6 +59,15 @@ def _construct_reconfiguraiton_wedge_coammand(self):
         reconf_msg.additional_data = bytes()
         return reconf_msg
 
+    def _construct_reconfiguraiton_latest_prunebale_block_coammand(self):
```
Comment (shgandhi) [inline tests/apollo/test_skvbc_reconfiguration.py:62]: Nit: change `_construct_reconfiguraiton_latest_prunebale_block_coammand` to `_construct_reconfiguration_latest_pruneable_block_command`.

File: tests/apollo/test_skvbc_reconfiguration.py [hunk 13]
```diff
@@ -68,6 +77,27 @@ def _construct_reconfiguraiton_wedge_status(self):
         reconf_msg.additional_data = bytes()
         return reconf_msg
 
+    def _construct_reconfiguraiton_prune_request(self, latest_pruneble_blocks):
```
Comment (shgandhi) [inline tests/apollo/test_skvbc_reconfiguration.py:80]: Nit: change `reconfiguraiton` to `reconfiguration`.

File: tests/apollo/test_skvbc_reconfiguration.py [hunk 14]
```diff
@@ -68,6 +77,27 @@ def _construct_reconfiguraiton_wedge_status(self):
...
+        prune_cmd = cmf_msgs.PruneRequest()
+        prune_cmd.sender = 1000
+        prune_cmd.latest_prunable_block = latest_pruneble_blocks
+        reconf_msg = cmf_msgs.ReconfigurationRequest()
+        reconf_msg.command = prune_cmd
+        reconf_msg.signature = bytes()
+        reconf_msg.additional_data = bytes()
+        return reconf_msg
+
+    def _construct_reconfiguraiton_prune_status_request(self):
```
Comment (shgandhi) [inline tests/apollo/test_skvbc_reconfiguration.py:90]: Nit: change `reconfiguraiton` to `reconfiguration`.

File: tests/apollo/test_skvbc_reconfiguration.py [hunk 15]
```diff
@@ -223,205 +253,138 @@ async def test_wedge_command_where_noops_should_be_sent_in_two_parts(self, bft_n
...
-             Sends a wedge command and check that the system stops from processing new requests.
-             Note that in this test we assume no failures and synchronized network.
-             The test does the following:
-             1. A client sends a wedge command
-             2. The client then sends a "Have you stopped" read only command such that each replica answers "I have stopped"
-             3. Apollo stops all replicas
-             4. Apollo starts all replicas
-             5. A client verifies that new write requests are being processed
-         """
+    async def test_get_latest_prunebable_block(self, bft_network):
```
Comment (shgandhi) [inline tests/apollo/test_skvbc_reconfiguration.py:258]: Nit: wherever applies, change `prunebable` to `pruneable`.

File: tests/apollo/util/concord_msgs.py [hunk 16]
```diff
@@ -758,6 +758,97 @@ def __eq__(self, other):
             return False
         return True
 
+class InstallCommand():
```
Comment (shgandhi) [inline tests/apollo/util/concord_msgs.py:766]: Nit: `UpgradeCommand` class and its usage should be removed in `concord_msgs.py`, since we are using `InstallCommand`.

File: reconfiguration/cmf/concord.cmf [hunk 17]
```diff
@@ -60,6 +60,19 @@ Msg GetVersionCommand 19 {
   bytes place_holder
 }
 
+Msg InstallCommand 20 {
+    string version
+}
+
+Msg InstallStatusCommand 21 {
+    string version
```
Comment (shgandhi) [inline reconfiguration/cmf/concord.cmf:69]: If we make the assumption that only one release version is being installed at a time,`version` can be made optional for the `InstallStatusCommand`. 
The main use-case: If we want to check at the operator if any installation is in progress before starting a new install, we should be able to run the install status command without a version argument.

**Review by shgandhi — COMMENTED** [review]
Thanks for the change @yontyon, @toly-kournik. 
I have reviewed the SW upgrade relevant parts. One recurring change required is the omission of `UpgradeCommand`, since it is replaced by `InstallCommand`. Other than that, I have added a few minor comments.

**Review by andrewjstone — COMMENTED** [review]
Given the size and importance of this PR, please give it a few days to soak before merging. 

I would like to review it thoroughly, and I think others would as well. I anticipate a thorough review will take at least a few hours.

File: bftclient/include/bftclient/base_types.h (hunk 1 above)
Comment (yontyon) [reply bftclient/include/bftclient/base_types.h:44]: MsgFlags contain flags that the client shouldn't be able to use (such as HAS_PRE_PROCESSED_FLAG).
I think this is why the original author (@andrewjstone ?) has separated between the two

File: bftengine/include/bftengine/Replica.hpp (hunk 6 above)
Comment (yontyon) [reply bftengine/include/bftengine/Replica.hpp:58]: The reason why IControlHandler is part of Replica.hpp is because it is part of the bftengine interfaces.
A user can load into the singleton a custom implementation that is relevant for his own application. Note that ControlHandler.hpp is part of the src directory of bftengine.

We decided to have an interface class to enlarge the flexibility.
This way, if we want to change in concord the way we handle the events in concordbft, we won't need to change the default implementation in cocordbft

File: bftengine/src/bftengine/ReplicaImp.hpp (hunk 10 above)
Comment (yontyon) [reply bftengine/src/bftengine/ReplicaImp.hpp:15]: it is included with RequestHandler.h

File: bftclient/include/bftclient/base_types.h (hunk 1 above)
Comment (andrewjstone) [reply bftclient/include/bftclient/base_types.h:44]: Basically yes. But it's also because the client doesn't link with the bftengine library on purpose. They are separate code bases.

**Review by toly-kournik — APPROVED** [review]
LGTM, Thanks

File: bftengine/include/bftengine/Replica.hpp (hunk 4 above)
Comment (yontyon) [reply bftengine/include/bftengine/Replica.hpp:56]: In the design we prefer to have a singleton on top of the complication of passing the reference throughout the whole stack

File: reconfiguration/src/dispatcher.cpp [hunk 18]
```diff
@@ -46,19 +46,28 @@ ReconfigurationResponse Dispatcher::dispatch(const ReconfigurationRequest& reque
...
+      rresp.response.emplace<concord::messages::GetVersionResponse>(response);
     } else if (holds_alternative<DownloadCommand>(request.command)) {
       LOG_INFO(getLogger(), "DownloadCommand");
       for (auto& handler : reconfig_handlers_)
         rresp.success &= handler->handle(std::get<DownloadCommand>(request.command));
-      if (rresp.success) ADDITIONAL_DATA(rresp, "Downloading");
     } else if (holds_alternative<UpgradeCommand>(request.command)) {
       LOG_INFO(getLogger(), "UpgradeCommand");
       for (auto& handler : reconfig_handlers_)
         rresp.success &= handler->handle(std::get<UpgradeCommand>(request.command));
```
[the remark this answers is not in the document]
Comment (yontyon) [reply reconfiguration/src/dispatcher.cpp]: done

File: reconfiguration/include/reconfiguration/reconfiguration_handler.hpp [hunk 19]
```diff
@@ -24,11 +24,11 @@ class ReconfigurationHandler : public IReconfigurationHandler {
  public:
   bool handle(const concord::messages::WedgeCommand&) override;
   bool handle(const concord::messages::WedgeStatusRequest&, concord::messages::WedgeStatusResponse&) override;
-  bool handle(const concord::messages::GetVersionCommand&) override;
+  bool handle(const concord::messages::GetVersionCommand&, concord::messages::GetVersionResponse&) override;
   bool handle(const concord::messages::DownloadCommand&) override;
   bool handle(const concord::messages::UpgradeCommand&) override;
```
[the remark this answers is not in the document]
Comment (yontyon) [reply reconfiguration/include/reconfiguration/reconfiguration_handler.hpp]: done

File: tests/apollo/test_skvbc_reconfiguration.py (hunk 12 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_reconfiguration.py]: done

File: tests/apollo/test_skvbc_reconfiguration.py (hunk 14 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_reconfiguration.py]: done

File: tests/apollo/test_skvbc_reconfiguration.py (hunk 13 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_reconfiguration.py]: done

File: tests/apollo/test_skvbc_reconfiguration.py (hunk 15 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_reconfiguration.py]: done

File: tests/apollo/util/concord_msgs.py (hunk 16 above)
Comment (yontyon) [reply tests/apollo/util/concord_msgs.py:739]: done

File: reconfiguration/cmf/concord.cmf (hunk 17 above)
Comment (yontyon) [reply reconfiguration/cmf/concord.cmf]: done

File: reconfiguration/include/reconfiguration/ireconfiguration.hpp [hunk 20]
```diff
@@ -27,11 +27,11 @@ class IReconfigurationHandler {
   // Message handlers
   virtual bool handle(const concord::messages::WedgeCommand&) = 0;
   virtual bool handle(const concord::messages::WedgeStatusRequest&, concord::messages::WedgeStatusResponse&) = 0;
-  virtual bool handle(const concord::messages::GetVersionCommand&) = 0;
+  virtual bool handle(const concord::messages::GetVersionCommand&, concord::messages::GetVersionResponse&) = 0;
   virtual bool handle(const concord::messages::DownloadCommand&) = 0;
   virtual bool handle(const concord::messages::UpgradeCommand&) = 0;
```
[the remark this answers is not in the document]
Comment (yontyon) [reply reconfiguration/include/reconfiguration/ireconfiguration.hpp]: done
